src/exec/use_pty/pipe.rs

```diff
@@ -142,78 +156,202 @@ impl<R: Read, W: Write> Buffer<R, W> {
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //           ▲           ▲
+            //           │           │
+            //         head         tail
+            // In the extreme case where head is as close as possible to tail, the sub-buffer will have a
+            // length of one. We can be sure that head and tail are not equal because the buffer is
+            // not empty.
+            //         ┌───┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │   │ M │ D │   │   │   │   │   │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //           ▲   ▲
+            //           │   │
+            //         head  tail
+            &mut self.buffer[self.head..self.tail]
         } else {
-            // Otherwise we just free the first `len` bytes of the busy section.
-            self.start += len;
-        }
+            // If head is greater than tail, the sub-buffer extends from head to the end of the
+            // buffer.
+            //                     ┌───────────┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │   │   │   │ M │ D │ D │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //       ▲               ▲
+            //       │               │
+            //     tail             head
+            //
+            // In the extreme case where head is the last position, the sub-buffer will have a
+            // length of one.
+            //                             ┌───┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │   │   │   │   │   │ M │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //       ▲                       ▲
+            //       │                       │
+            //     tail                     head
+            // On the other hand if tail is as close as it can be to head, the buffer is full
+            //             ┌───────────────────┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │ D │ M │ D │ D │ D │ D │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //           ▲   ▲
+            //           │   │
+            //         tail head
+            // We know they cannot be equal because the buffer is not empty.
+            &mut self.buffer[self.head..]
+        };
+
+        // Remove `len` bytes from the buffer and write them.
+        let len = write.write(buffer)?;
 
-        // If we wrote something, the buffer is not full anymore and we can resume reading.
         if len > 0 {
+            // We update head so it becomes the first position with data again.
+            self.head += len;
+            // However, we must be sure that tail doesn't go past the last position.
+            self.head %= self.buffer.len();
+            // If we actually removed something, the buffer is not full anymore and we can resume
+            // reading.
             self.read_handle.resume(registry);
         }
```

Comment:
similar situation here: in the `len == buffer.len()` case you may be able to write out more data if there's more data in the front of the ring buffer (`tail < head` case)